deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,239 @@
+            raise ValueError(
+                "We couldn't find the workspace to publish to in your environment. "
+                "Please run 'deepset-cloud login' and follow the instructions."
+            )
+
+        if isinstance(config, IndexConfig):
+            logger.debug(f"Publishing index to workspace {DEFAULT_WORKSPACE_NAME}")
+            await self._publish_index(pipeline, config)
+        else:
+            logger.debug(f"Publishing pipeline to workspace {DEFAULT_WORKSPACE_NAME}")
+            await self._publish_pipeline(pipeline, config)
+
+    async def _publish_index(self, pipeline: PipelineProtocol, config: IndexConfig) -> None:
+        """Publish an index pipeline to deepset Cloud.
+
+        :param pipeline: The Haystack pipeline to publish
+        :param config: Configuration for publishing
+        """
+        pipeline_yaml = self._create_config_yaml(pipeline, config)
+        response = await self._api.post(
+            workspace_name=DEFAULT_WORKSPACE_NAME,
+            endpoint="indexes",
+            json={"name": config.name, "config_yaml": pipeline_yaml},
+        )
+        response.raise_for_status()
+        if response.status_code == 201:
+            logger.debug(f"Index {config.name} successfully created")
+
+    async def _publish_pipeline(self, pipeline: PipelineProtocol, config: PipelineConfig) -> None:
+        """Publish a pipeline to deepset Cloud.
+
+        :param pipeline: The pipeline to publish
+        :param config: Configuration for publishing
+        """
+        logger.debug(f"Publishing pipeline {config.name}")
+        pipeline_yaml = self._create_config_yaml(pipeline, config)
+        response = await self._api.post(
+            workspace_name=DEFAULT_WORKSPACE_NAME,
+            endpoint="pipelines",
+            json={"name": config.name, "query_yaml": pipeline_yaml},
+        )
+        response.raise_for_status()
+        if response.status_code == 201:
+            logger.debug(f"Pipeline {config.name} successfully created")
+
+    def _create_config_yaml(self, pipeline: PipelineProtocol, config: IndexConfig | PipelineConfig) -> str:
+        """Create the config YAML string.
+
+        :param pipeline: The Haystack pipeline to convert to YAML
+        :param config: Configuration for publishing
+        :return: Complete YAML string with inputs and outputs
+        """
+        # Parse the pipeline YAML
+        pipeline_yaml = self._yaml.load(pipeline.dumps())
+
+        # Add inputs and outputs
+        if config.inputs and (converted_inputs := self._convert_to_yaml_dict(config.inputs)):
+            pipeline_yaml["inputs"] = converted_inputs
+        if config.outputs and (converted_outputs := self._convert_to_yaml_dict(config.outputs)):
+            pipeline_yaml["outputs"] = converted_outputs
+
+        # Convert back to string
+        string_stream = StringIO()
+        self._yaml.dump(pipeline_yaml, string_stream)
+        return string_stream.getvalue()
```

Comment:
nit: could also be some `from_haystack_pipeline` within the model, right ? 

I think its also fine to keep it like this for now. Was just thinking it is easier readable if the models themselfes parse/convert themselfes instead of having data wrangling helper functions in the service. Might even simply doable by defining a "PipelineConfig" and extending it as Index + Pipeline and adding the create_config_yaml to the PipelineConfig